### PR TITLE
feat(levm): environment opcodes

### LINKED ISSUE #509: Opcode: SELFBALANCE

### LINKED ISSUE #512: Opcode: ADDRESS

### LINKED ISSUE #513: Opcode: BALANCE

### LINKED ISSUE #514: Opcode: ORIGIN

### LINKED ISSUE #515: Opcode: CALLER

### PR DESCRIPTION
**Pull request by lazcanoluca**
**Motivation**

To implement some missing environment opcodes.

**Description**

Closes #515 
Closes #514 
Closes #513 
Closes #512 
Closes #509

### CODE REVIEW

File: crates/vm/levm/src/opcode_handlers/block.rs [hunk 1]
```diff
@@ -206,33 +207,75 @@ impl VM {
...
         &mut self,
-        _current_call_frame: &mut CallFrame,
+        current_call_frame: &mut CallFrame,
     ) -> Result<OpcodeSuccess, VMError> {
-        unimplemented!();
+        if self.env.consumed_gas + gas_cost::ADDRESS > self.env.gas_limit {
+            return Err(VMError::OutOfGas);
+        }
+
+        let addr = current_call_frame.code_address;
```
Comment (iinaki) [inline crates/vm/levm/src/opcode_handlers/block.rs:216]: This works fine in most cases, but not if we are currently in a `DELEGATECALL`, then the address would be msg.sender, not the address of the called contract.

**Review by iinaki — COMMENTED** [review]
Just noticed this little detail, the rest looks good to me!

File: crates/vm/levm/src/opcode_handlers/block.rs (hunk 1 above)
Comment (lazcanoluca) [reply crates/vm/levm/src/opcode_handlers/block.rs]: Fixed!

File: crates/vm/levm/src/vm.rs [hunk 2]
```diff
@@ -110,6 +110,11 @@ impl Db {
     pub fn add_account(&mut self, address: Address, account: Account) {
         self.accounts.insert(address, account);
     }
+
+    pub fn with_account(mut self, address: Address, account: Account) -> Self {
+        self.add_account(address, account);
+        self
+    }
```
[the remark this answers is not in the document]
Comment (lazcanoluca) [reply crates/vm/levm/src/vm.rs]: You're right! Removed

**Review by iinaki — APPROVED** [review]

**Review by juanimedone — APPROVED** [review]

**Pull request merged**